text/0240-Sampling_Configuration.md

```diff
@@ -0,0 +1,163 @@
+
+In contrast, the existing configuration schemas, such as [Jaeger sampling](https://www.jaegertracing.io/docs/1.50/sampling/#file-based-sampling-configuration), or Agent [OTEP 225 - Configuration proposal](https://github.com/open-telemetry/oteps/pull/225) address sampling configuration with a limited known set of samplers only.
+
+## Use cases
+
+- I want to use one of the samplers from the `opentelemetry-java-contrib` repository, but I do not want to build my own agent extension. I prefer to download one or more jarfiles containing the samplers and configure their use without writing any additional code.
+- I want to apply a sampling strategy that combines different samplers depending on the span attributes, such as the URL of the incoming request, and I expect to update the configuration frequently, so I prefer that it is file-based (rather than hardcoded), and better yet, applied dynamically
+ - I want to write my own sampler with some unique logic, but I want to focus entirely on the sampling algorithm and avoid writing any boilerplate code for instantiating, configuring, and wrapping it up as an agent extension
+
+## The basics
+
+It is assumed that the sampling configuration will be a YAML document, in most cases available as a file. Remote configuration remains an option, as well as dynamic changes to the configuration.
+
+The configuration file will contain a definition of the sampler to use:
+
+```yaml
+---
+sampler:
+   SAMPLER
+```
+Additional information could be placed there as well. For example, for Java, there could be a location of a jarfile containing any non-standard samplers which have been configured.
+
+A SAMPLER is described by a structure with two fields:
+```yaml
+  samplerType: TYPE
+  parameters:    # an optional list of parameters for the sampler
+   - PARAM1
+   - PARAM2
+   ...
+   - PARAMn
+```
+The mechanism to map the sampler TYPE to the sampler implementation will be platform specific. For Java, the TYPE can be a simple class name, or a part of class name, and the implementing class can be found using reflection. Specifying a fully qualified class name should be an option.
+
+There will be a need for each supported sampler to have a documented canonical way of instantiation or initialization, which takes a known list of parameters. The order of the parameters specified in the configuration file will have to match exactly that list.
+
+A sampler can be passed as an argument for another sampler:
+```yaml
+   samplerType: ParentBased
+   parameters:
+    - samplerType: TraceIdRatioBased  # for root spans
+      parameters:
+       - 0.75
+    - samplerType: AlwaysOn    # remote parent sampled
+    - samplerType: AlwaysOff   # remote parent not sampled
+    - samplerType: AlwaysOn    # local parent sampled
+    - samplerType: AlwaysOff   # local parent not sampled
+```
```

Comment:
this example is confusing - ParentBased takes only one base sampler as argument, does it not?

Reply:
No, it takes five arguments. Perhaps you got it confused with `ParentPasedSamplerBuilder`, which takes one.

Reply:
if those are 5 distinct base samplers, why not name them accordingly in yaml instead of 
`-samplerType`?

Reply:
Because that would require the parser to know that ParentBasedSampler takes 5 samplers as arguments (and not Strings, for example).

Reply:
parser already needs to know that. If this was a different sampler like TraceIdRatioBased it would be one parameter with numeric value - how is that knowledge different from knowing about 5 args?

Reply:
YAML parser natively provides Numbers, Strings, an possibly nested Lists and Maps (with Strings as keys). Detection of known higher-order types such as Samplers, Predicates and Rules should be easy. Then the parser will build an argument list for a parser to instantiate. Instantiation failure will indicate an error in the configuration.
So it is necessary to distinguish a case where a sampler takes 5 Samplers as arguments from the case where it takes a list of Samplers. This difference is reflected in the proposed YAML syntax.

Reply:
This does not explain why parameters cannot be named according to their semantics in the specific sampler.

Reply:
A list is needed because the order of the arguments is essential. A map of parameter names to values would be unordered and the parser would have to know the order of the parameter names required for sampler instantiation.

Reply:
Sorry, but I don't see convincing arguments here why config needs to be so cryptic, with unnamed positional arguments, which is very user-unfriendly. @jack-berg just [pointed to an alternative](https://github.com/open-telemetry/opentelemetry-configuration/blob/main/examples/kitchen-sink.yaml#L343-L349) that does not sacrifice readability for some implementation details

Reply:
Modified the schema to match exactly the syntax proposed by [OTEP 225](https://github.com/open-telemetry/oteps/pull/225) while retaining additional capabilities.